eth/api_backend.go

```diff
@@ -53,7 +56,28 @@ func (b *EthAPIBackend) CurrentBlock() *types.Block {
 
 func (b *EthAPIBackend) SetHead(number uint64) {
 	b.eth.protocolManager.downloader.Cancel()
+
+	// Note: Suddenly changing the state of the blockchain can break other parts
+	// of the codebase that are holding on to some state, in some cases causing
+	// crashes. In order to prevent that, we need to stop and reset the miner,
+	// txpool, and consensus engine.
+	wasMining := b.eth.miner.Mining()
+	if wasMining {
+		b.eth.miner.Stop()
+	}
+	b.eth.txPool.Stop()
 	b.eth.blockchain.SetHead(number)
+	b.eth.txPool = core.NewTxPool(b.eth.config.TxPool, b.eth.chainConfig, b.eth.blockchain)
+	if b.eth.chainConfig.Clique != nil {
```

Comment:
We use a single node network for testing with Clique as the consensus engine. I'm not sure how to handle Ethash or other consensus engines here. Is that something we need to consider?